### PR TITLE
Add cutback_factor_at_failure in TimeStepper

### LINKED ISSUE #13873: Allow for user input for time step size cutback factor at failure
## Reason
Currently, if a time step fails to converge, it is hard-coded to have the time step cutback into half. This is not always optimal, and should let user decide how to cut the time step size.

## Design
Add an input parameter, 'cutback_factor_at_failuer' at the TimeStepper base class level to accept user-specified such a factor.

## Impact
No.

### PR DESCRIPTION
**Pull request by lingzou**
This new input parameter accepts user input and replaces the hard-coded factor 0.5. 

closes #13873

### CODE REVIEW

**Review by permcody — COMMENTED** [review]
Looks good - Can you throw in a quick test? There should be a few that test failures already. Just override this new parameter so that you can see it in action.

File: framework/src/timesteppers/TimeStepper.C [hunk 1]
```diff
@@ -44,6 +47,7 @@ TimeStepper::TimeStepper(const InputParameters & parameters)
     _timestep_tolerance(_executioner.timestepTol()),
     _verbose(_executioner.verbose()),
     _converged(true),
+    _cutback_factor_at_failure(getParam<Real>("cutback_factor_at_failuer")),
```
Comment (dschwen) [inline framework/src/timesteppers/TimeStepper.C]:
```suggestion
    _cutback_factor_at_failure(getParam<Real>("cutback_factor_at_failure")),
```

**Review by dschwen — CHANGES_REQUESTED** [review]
Typo

File: framework/include/timesteppers/TimeStepper.h [hunk 2]
```diff
@@ -138,6 +138,9 @@ class TimeStepper : public MooseObject, public Restartable, public ScalarCouplea
   /// Whether or not the previous solve converged.
   bool _converged;
 
+  /// Cutback factor if a time step fails to converge
+  const Real & _cutback_factor_at_failure;
```
Comment (dschwen) [inline framework/include/timesteppers/TimeStepper.h]:
```suggestion
  const Real _cutback_factor_at_failure;
```

This should be captured by value.

**Review by dschwen — CHANGES_REQUESTED** [review]

File: framework/include/timesteppers/TimeStepper.h (hunk 2 above)
Comment (permcody) [reply framework/include/timesteppers/TimeStepper.h]: Well that depends. If you ever want to control this parameter from the Control system, capturing by reference is the correct way.

File: framework/src/timesteppers/TimeStepper.C (hunk 1 above)
Comment (permcody) [reply framework/src/timesteppers/TimeStepper.C]: Good catch! That's why we do testing.

Comment (lingzou) [reply framework/src/timesteppers/TimeStepper.C]: Thanks, fixed at the updated commit.

File: framework/include/timesteppers/TimeStepper.h (hunk 2 above)
Comment (lingzou) [reply framework/include/timesteppers/TimeStepper.h]: I left it as-is.

**Comment by lingzou** [comment]
Two tests were added to test this new feature:
time_steppers/cutback_factor_at_failure.constant_dt_cutback
time_steppers/cutback_factor_at_failure.function_dt_cutback

**Comment by permcody** [comment]
This will fail the documentation target because every test has to have either a unique requirement, or you can add another level of nesting and have a single requirement with "details" about what makes each test unique. See the example at the very bottom of this page.
https://www.mooseframework.org/python/MooseDocs/generate.html

Thanks for adding these tests!

**Comment by lingzou** [comment]
@permcody Thanks for the guidance, I have updated the test requirements, and hope they will go through this time.

**Comment by permcody** [comment]
@lingzou - I think I've been doing this too long but the cutback factor should actually be range restricted. The way you have it now, cut back factors less than zero or greater than or equal to one would be really bad!

**Comment by permcody** [comment]
Actually no - negative is ok. It'll immediately cut back to min_dt... Still though

File: framework/include/timesteppers/TimeStepper.h (hunk 2 above)
Comment (dschwen) [reply framework/include/timesteppers/TimeStepper.h]: That would require a `params.declareControllable("cutback_factor_at_failure");`. Either add that, or get rid of the reference.

**Review by dschwen — CHANGES_REQUESTED** [review]

**Review by permcody — APPROVED** [review]

File: framework/src/timesteppers/TimeStepper.C [hunk 3]
```diff
@@ -184,10 +190,10 @@ TimeStepper::computeFailedDT()
   if (_dt <= _dt_min)
     mooseError("Solve failed and timestep already at or below dtmin, cannot continue!");
 
-  // cut the time step in a half
-  if (0.5 * _dt >= _dt_min)
-    return 0.5 * _dt;
-  else // (0.5 * _current_dt < _dt_min)
+  // cut the time step
+  if (_cutback_factor_at_failure * _dt >= _dt_min)
```
Comment (dschwen) [inline framework/src/timesteppers/TimeStepper.C]: This if else block could be replaced by this single line:
```
return std::max(_cutback_factor_at_failure * _dt, _dt_min);
```

Comment (permcody) [reply framework/src/timesteppers/TimeStepper.C]: yeah it could be, but at some point, we probably should just accept working, valid code :)

Comment (dschwen) [reply framework/src/timesteppers/TimeStepper.C]: Ha, yeah, that should not hold up this PR. I've seen this antipattern before though, so I thought I point it out.

**Review by dschwen — APPROVED** [review]

**Pull request merged**